t/lib/t/MusicBrainz/DataStore/RedisMulti.pm

```diff
@@ -0,0 +1,127 @@
+use MusicBrainz::DataStore::Redis;
+use MusicBrainz::DataStore::RedisMulti;
+use DBDefs;
+
+=head2 Test description
+
+This test checks basic tasks for the RedisMulti store (including adding,
+deleting and expiring keys), which distributes queries to multiple underlying
+stores. Consistency among all the stores is checked after each operation.
+
+=cut
+
+# Initialize tests
+my $args1 = DBDefs->DATASTORE_REDIS_ARGS;
+$args1->{database} = $args1->{test_database};
+
+my $args2 = DBDefs->DATASTORE_REDIS_ARGS;
+$args2->{database} = $args2->{test_database} + 1;
+
+my $redis1 = MusicBrainz::DataStore::Redis->new($args1);
+my $redis2 = MusicBrainz::DataStore::Redis->new($args2);
+
+my $redis_multi = MusicBrainz::DataStore::RedisMulti->new(
+    _redis_instances => [$redis1, $redis2],
+);
+
+# This doesn't test RedisMulti behavior specifically, but is a prerequisite
+# for the rest of these tests to make sense.
+test 'Databases are separate' => sub {
+    $redis1->set('kx1', 'vx1');
+    $redis2->set('kx2', 'vx2');
+
+    is($redis1->get('kx1'), 'vx1', 'Expected string is in database 1');
+    is($redis2->exists('kx1'), 0, 'String from database 1 is not in database 2');
+
+    is($redis2->get('kx2'), 'vx2', 'Expected string is in database 2');
+    is($redis1->exists('kx2'), 0, 'String from database 2 is not in database 1');
+
+    $redis1->clear;
+    $redis2->clear;
+};
+
+test 'Test key setting/retrieving' => sub {
+    is($redis_multi->get('does-not-exist'), undef, 'Non-existent key returns undef');
+
+    $redis_multi->set('string', 'Esperándote');
+    is($redis_multi->get('string'), 'Esperándote', 'Retrieved expected string');
+    is($redis1->get('string'), 'Esperándote', 'Expected string is in database 1');
+    is($redis2->get('string'), 'Esperándote', 'Expected string is in database 2');
+
+    $redis1->set('string-1-only', '1-only');
+    $redis1->set('string-2-only', '2-only');
+    is($redis_multi->get('string-1-only'), '1-only', 'Retrieved string that was only set in database 1');
+    is($redis_multi->get('string-2-only'), '2-only', 'Retrieved string that was only set in database 2');
+
+    is($redis_multi->exists('does-not-exist'), 0, 'exists returns 0 for non-existent key');
+    is($redis1->exists('does-not-exist'), 0, 'exists returns 0 for non-existent key in database 1');
+    is($redis2->exists('does-not-exist'), 0, 'exists returns 0 for non-existent key in database 2');
+
+    is($redis_multi->exists('string'), 1, 'exists returns 1 for existing key');
+    is($redis1->exists('string'), 1, 'exists returns 1 for existing key in database 1');
+    is($redis2->exists('string'), 1, 'exists returns 1 for existing key in database 2');
+
+    $redis_multi->delete('string');
+    is($redis_multi->exists('string'), 0, 'exists returns 0 for deleted key');
+    is($redis1->exists('string'), 0, 'exists returns 0 for deleted key in database 1');
+    is($redis2->exists('string'), 0, 'exists returns 0 for deleted key in database 2');
```

Comment:
What should happen if you delete from `$redis1` but not `$redis2`? Should we have a test for that case?

Reply:
It'd only be deleted from `$redis1`,  but any servers with `RedisMulti` enabled would still see the key if it exists in `$redis2`.

This shouldn't really happen in practice, because we'd be deploying the `RedisMulti` configuration on all servers at once.  But there is a minor race condition where a key is set on a server with a `RedisMulti` configuration, and then deleted on a server that doesn't have the new configuration yet (because we're in the middle of deploying it).  The only situation I can see this happening is if a user logs in and then logs out; this would have to be perfectly timed and forwarded to the right servers.  In that case there may be a stale session key on the new Redis instance, but it would expire within some hours anyway.

Test-wise we do have the "string-1-only"/"string-2-only" ones here, which check that keys set in only one of the servers will be retrieved through `$redis_multi` (doesn't matter if they're deleted first).